streams/src/test/java/org/apache/kafka/streams/integration/KStreamRepartitionIntegrationTest.java

```diff
@@ -0,0 +1,807 @@
+
+        CLUSTER.createTopic(topicB, 1, 1);
+
+        final StreamsBuilder builder = new StreamsBuilder();
+
+        final Repartitioned<Integer, String> inputTopicRepartitioned = Repartitioned
+            .<Integer, String>as(inputTopicRepartitionedName)
+            .withNumberOfPartitions(inputTopicNumberOfPartitions);
+
+        final Repartitioned<Integer, String> topicBRepartitioned = Repartitioned
+            .<Integer, String>as(topicBRepartitionedName)
+            .withNumberOfPartitions(topicBNumberOfPartitions);
+
+        final KStream<Integer, String> topicBStream = builder
+            .stream(topicB, Consumed.with(Serdes.Integer(), Serdes.String()))
+            .repartition(topicBRepartitioned);
+
+        builder.stream(inputTopic, Consumed.with(Serdes.Integer(), Serdes.String()))
+               .repartition(inputTopicRepartitioned)
+               .join(topicBStream, (value1, value2) -> value2, JoinWindows.of(Duration.ofSeconds(10)))
+               .to(outputTopic);
+
+        startStreams(builder, REBALANCING, ERROR, (t, e) -> throwable.set(e));
+
+        final Map<String, Integer> repartitionTopicsWithNumOfPartitions = Utils.mkMap(
+            Utils.mkEntry(toRepartitionTopicName(topicBRepartitionedName), topicBNumberOfPartitions),
+            Utils.mkEntry(toRepartitionTopicName(inputTopicRepartitionedName), inputTopicNumberOfPartitions)
+        );
+
+        final String expectedErrorMessage = String.format("Following topics do not have the same " +
+                                                          "number of partitions: [%s]",
+                                                          new TreeMap<>(repartitionTopicsWithNumOfPartitions));
+
+        assertNotNull(throwable.get());
+        assertTrue(throwable.get().getMessage().contains(expectedErrorMessage));
+    }
+
+    @Test
+    public void shouldThrowAnExceptionWhenNumberOfPartitionsOfRepartitionOperationDoNotMatchSourceTopicWhenJoining() throws InterruptedException {
+        final String topicBMapperName = "topic-b-mapper";
+        final String topicB = "topic-b-" + TEST_NUM.get();
+        final int topicBNumberOfPartitions = 6;
+        final String inputTopicRepartitionName = "join-repartition-test";
+        final AtomicReference<Throwable> expectedThrowable = new AtomicReference<>();
+        final int inputTopicRepartitionedNumOfPartitions = 2;
+
+        CLUSTER.createTopic(topicB, topicBNumberOfPartitions, 1);
+
+        final StreamsBuilder builder = new StreamsBuilder();
+
+        final Repartitioned<Integer, String> inputTopicRepartitioned = Repartitioned
+            .<Integer, String>as(inputTopicRepartitionName)
+            .withNumberOfPartitions(inputTopicRepartitionedNumOfPartitions);
+
+        final KStream<Integer, String> topicBStream = builder
+            .stream(topicB, Consumed.with(Serdes.Integer(), Serdes.String()))
+            .map(KeyValue::new, Named.as(topicBMapperName));
```

Comment:
I am wondering, if we should really throw an exception for this case? Why do we not create this repartition topic with the same number of partitions as specified on the second input topic instead?

IIRC, for the following case we would also adjust the number of partitions:
```
topicWith2partitions = builder.stream(...).map(...);
topicWith4partitions = builder.stream(...);
topicWith2Partitions.join(topicWith4partitions,...)
```
This code above is very similar to:
```
topicWith2partitions = builder.stream(...).map(...);
topicWith4partitions = builder.stream("topicWith1Partitions").repartition(/*set 4 partitions*/);
topicWith2Partitions.join(topicWith4partitions,...)
```

In both cases, the number of partitions of one topic is fixed, while the second one has key-changing operation and thus we can just create a reparition topic that matches the number of partitions of the first topics?

Reply:
Interesting point. I guess in that case idea would be if in the `copartitionGroup` there's one `ImmutableRepartitionTopicConfig` and rest are repartition topics, we will enforce number of partitions from `ImmutableRepartitionTopicConfig`. In cases when there're more than one `ImmutableRepartitionTopicConfig` we would still throw an exception. Does this make sense?

Reply:
Hi @mjsax this is now implemented.